core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java

```diff
@@ -146,27 +185,12 @@ boolean shardCanBeDeleted(ClusterState state, IndexShardRoutingTable indexShardR
 
         for (ShardRouting shardRouting : indexShardRoutingTable) {
             // be conservative here, check on started, not even active
-            if (!shardRouting.started()) {
+            if (shardRouting.started() == false) {
                 return false;
             }
 
-            // if the allocated or relocation node id doesn't exists in the cluster state  it may be a stale node,
-            // make sure we don't do anything with this until the routing table has properly been rerouted to reflect
-            // the fact that the node does not exists
-            DiscoveryNode node = state.nodes().get(shardRouting.currentNodeId());
```

Comment:
>  I'll investigate where to put the assertion (it's complementary to this PR).

++

re the speed - I run IndicesClusterStateServiceRandomUpdatesTests with this patch in it (i.e., 100 times and the test it self iterates too). The performance difference is negligible (46s vs 44s - which is just random noise). I don't think we need to worry much about performance here.

```
diff --git a/core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java b/core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java
index 074b4a5..2016dad 100644
--- a/core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java
+++ b/core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java
@@ -19,6 +19,7 @@

 package org.elasticsearch.indices.cluster;

+import com.carrotsearch.randomizedtesting.annotations.Repeat;
 import org.apache.logging.log4j.message.ParameterizedMessage;
 import org.elasticsearch.Version;
 import org.elasticsearch.action.admin.cluster.reroute.ClusterRerouteRequest;
@@ -71,6 +72,7 @@ import static org.elasticsearch.cluster.routing.ShardRoutingState.INITIALIZING;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;

+@Repeat(iterations = 100)
 public class IndicesClusterStateServiceRandomUpdatesTests extends AbstractIndicesClusterStateServiceTestCase {

     private final ClusterStateChanges cluster = new ClusterStateChanges();
@@ -341,6 +343,15 @@ public class IndicesClusterStateServiceRandomUpdatesTests extends AbstractIndice

         // TODO: go masterless?

+        for (ShardRouting shard: state.routingTable().allShards()) {
+            if (shard.assignedToNode()) {
+                assert state.nodes().get(shard.currentNodeId()) != null;
+            }
+            if (shard.relocating()) {
+                assert state.nodes().get(shard.relocatingNodeId()) != null;
+            }
+        }
+
         return state;
     }
```